logd/server/collector.go

```diff
@@ -0,0 +1,308 @@
+						lc.agent.Fail(fmt.Errorf("removing old events: %w", err))
+					}
+				}()
+			}
+		}
+	})
+
+	lc.eg.Go(func() error {
+		return lc.agent.Run(ctx)
+	})
+
+	return lc.eg.Wait()
+}
+
+func (lc *LogCollector) removeOldEvents(ctx context.Context) error {
+	lc.debugf("removing old events")
+	for logName := range lc.state.Logs {
+		log := lc.store.GetLog(logName)
+		if err := log.RemoveOldEvents(ctx); err != nil {
+			return fmt.Errorf("removing %q events: %w", logName, err)
+		}
+	}
+	lc.debugf("removed old events")
+	return nil
+}
+
+func validLogName(s string) bool {
+	return s != "" // TODO: More validation. Cannot have internal null bytes.
+}
+
+func (lc *LogCollector) AddLog(ctx context.Context, input *api.AddLogInput) (output *api.AddLogOutput, err error) {
+	err = lc.agent.Send(func() error {
+		output, err = lc.addLog(ctx, input)
+		return err
+	})
+	return
+}
+
+func (lc *LogCollector) addLog(ctx context.Context, input *api.AddLogInput) (*api.AddLogOutput, error) {
+	if !validLogName(input.Name) {
+		return nil, fmt.Errorf("invalid log name: %q", input.Name)
+	}
+	if input.Source == "" {
+		return nil, errors.New("log source path is required")
+	}
+	if lc.state.Logs == nil {
+		lc.state.Logs = make(map[string]*logState)
+	}
+	if lc.state.Logs[input.Name] != nil {
+		return nil, fmt.Errorf("already have log %q", input.Name)
+	}
+	logState := &logState{
+		Source: input.Source,
+	}
+	lc.state.Logs[input.Name] = logState
+	if err := lc.saveState(); err != nil {
+		return nil, fmt.Errorf("saving new log source: %w", err)
+	}
+
+	lc.startWorker(ctx, input.Name, logState)
+	return &api.AddLogOutput{}, nil
+}
+
+func (lc *LogCollector) startWorker(ctx context.Context, logName string, state *logState) {
+	ctx, stop := context.WithCancel(ctx)
+	wkr := &collectorWorker{
+		Worker: Worker{
+			Source: state.Source,
+			Sink:   lc.store.GetLog(logName),
+			Debug:  lc.debug,
+		},
+		stop: stop,
+	}
+	lc.workers[logName] = wkr
+
+	lc.eg.Go(func() error {
+		if err := wkr.Run(ctx); err != nil {
+			return fmt.Errorf("worker %q run error: %w", logName, err)
+		}
+		return nil
+	})
+}
+
+func (lc *LogCollector) RemoveLog(ctx context.Context, input *api.RemoveLogInput) (output *api.RemoveLogOutput, err error) {
+	err = lc.agent.Send(func() error {
+		output, err = lc.removeLog(ctx, input)
+		return err
+	})
+	return
+}
+
+func (lc *LogCollector) removeLog(ctx context.Context, input *api.RemoveLogInput) (*api.RemoveLogOutput, error) {
+	delete(lc.state.Logs, input.Name)
+	lc.stopWorker(input.Name)
+	if err := lc.saveState(); err != nil {
+		return nil, fmt.Errorf("saving log source removal: %w", err)
+	}
+	return &api.RemoveLogOutput{}, nil
+}
+
+func (lc *LogCollector) stopWorker(logName string) {
+	wkr := lc.workers[logName]
+	if wkr == nil {
+		return
+	}
+	wkr.stop()
+}
+
+func (lc *LogCollector) DescribeLogs(ctx context.Context, input *api.DescribeLogsInput) (output *api.DescribeLogsOutput, err error) {
+	err = lc.agent.Send(func() error {
```

Comment:
Okay, this pattern is kinda nice :)

Reply:
It would be nice to be able to use capabilities of some sort to guard access to describeLogs outside of agent ;)

Reply:
It took me a long while to get to this pattern! I think this is the least-bad, memory-model/concurrency-safe approach that doesn't require reflection.

Reply:
RE: Capabilities. Yeah, absolutely. I have some ideas on that front. It's starting to become relevant as I get moving on multiple-workspaces too. Will discuss in that PR when it comes up.